spark/src/main/scala/geotrellis/spark/io/hadoop/HadoopCollectionReader.scala

```diff
@@ -0,0 +1,93 @@
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.io._
+import org.apache.hadoop.fs.Path
+import scalaz.std.vector._
+import scalaz.concurrent.Task
+import scalaz.stream.{Process, nondeterminism}
+import com.typesafe.config.ConfigFactory
+
+import java.util.concurrent.Executors
+
+class HadoopCollectionReader(maxOpenFiles: Int) {
+  val readers = new LRUCache[Path, MapFile.Reader](maxOpenFiles.toLong, {x => 1l}) {
+    override def evicted(reader: MapFile.Reader) = reader.close()
+  }
+
+  def read[
+    K: AvroRecordCodec: Boundable,
+    V: AvroRecordCodec
+  ](path: Path,
+    conf: Configuration,
+    queryKeyBounds: Seq[KeyBounds[K]],
+    decomposeBounds: KeyBounds[K] => Seq[(Long, Long)],
+    indexFilterOnly: Boolean,
+    writerSchema: Option[Schema] = None,
+    numPartitions: Option[Int] = None,
+    threads: Int = ConfigFactory.load().getInt("geotrellis.hadoop.threads.collection.read")): Seq[(K, V)] = {
+    if (queryKeyBounds.isEmpty) return Seq.empty[(K, V)]
+
+    val includeKey = (key: K) => KeyBounds.includeKey(queryKeyBounds, key)
+    val indexRanges = queryKeyBounds.flatMap(decomposeBounds).toArray
+
+    val bins = IndexRanges.bin(indexRanges, numPartitions.getOrElse(CollectionLayerReader.defaultNumPartitions)).toVector.map(_.toIterator)
+
+    val codec = KeyValueRecordCodec[K, V]
+
+    val pathRanges: Vector[(Path, Long, Long)] =
+      FilterMapFileInputFormat.layerRanges(path, conf)
+
+    val pool = Executors.newFixedThreadPool(threads)
+
+    val result = bins flatMap { partition =>
+      val range: Process[Task, Iterator[Long]] = Process.unfold(partition) { iter =>
+        if (iter.hasNext) {
+          val (start, end) = iter.next()
+          Some((start to end).toIterator, iter)
+        }
+        else None
+      }
+
+      val read: Iterator[Long] => Process[Task, Vector[(K, V)]] = { iterator =>
+          Process.unfold(iterator) { iter =>
+            if (iter.hasNext) {
+              val index = iter.next()
+              val valueWritable = pathRanges
+                .find { row => index >= row._2 && index <= row._3 }
+                .map { case (p, _, _) => readers.getOrInsert(p, new MapFile.Reader(p, conf)) }
+                .getOrElse(throw new LayerIOError(s"Index ${index} not found."))
+                .get(new LongWritable(index), new BytesWritable())
+                .asInstanceOf[BytesWritable]
+
+              if (valueWritable == null) Some(Vector(), iter)
+              else {
+                val items = AvroEncoder.fromBinary(writerSchema.getOrElse(codec.schema), valueWritable.getBytes)(codec)
+                if (indexFilterOnly) Some(items, iter)
+                else Some(items.filter { row => includeKey(row._1) }, iter)
+              }
+            } else {
+              None
+            }
+          }
+      }
+
+      nondeterminism.njoin(maxOpen = threads, maxQueued = threads) { range map read }.runFoldMap(identity).unsafePerformSync
```

Comment:
I can see how that would mess with the caching though, if you had multiple async calls n to a Collection reader. If that's the case it almost makes it seem like CollectionsReader needs to be closable, as it will invariably represent some resources.

Main use case for these is to be called from endpoints to work on "small" areas, so async `read` calls are totally in scope of design.